Eliminate copy assignment of BigInteger

ts2do opened the pull request:
Remove constructors from BigInteger which end up causing copy assignments that copy 464 bytes to the target variable. Instead, default and nontrivial assignments will all be done using SetZero, SetUInt32, SetUInt64, and SetValue to only set the relevant blocks.

Unrelated: I'm not sure why I always encounter problems trying to build! This is a relatively clean install of Windows and Visual Studio 2019, and I'm seeing this issue:
```C:\Program Files\dotnet\sdk\3.0.100\Microsoft.Common.CurrentVersion.targets(1175,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.6 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [C:\dev\dotnet-runtime\tools-local\tasks\installer.tasks\installer.tasks.csproj]```
I have installed the .NET Framework 4.6 targeting pack using Visual Studio Installer, but it still gives the same message.

ts2do committed:
Eliminate copy assignment of BigInteger

Remove constructors from BigInteger which end up causing copy assignments that copy 464 bytes to the target variable. Instead, default and nontrivial assignments will all be done using SetZero, SetUInt32, SetUInt64, and SetValue to only set the relevant blocks.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:
```diff
@@ -1047,9 +1028,18 @@ public void Multiply(uint value)
+                    case 1:
+                        Multiply(value._blocks[0]);
+                        break;
+                    default:
```
This case is missing the `break` statement

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:
```diff
@@ -342,7 +325,6 @@ public static void Add(ref BigInteger lhs, ref BigInteger rhs, out BigInteger re
                 int smallLength = small._length;
 
                 // The output will be at least as long as the largest input
-                result = new BigInteger(0);
```
Doesn't `result` still need to be zero initialized or is this bypassed because `_blocks` is a fixed sized buffer?

ts2do replied:
Correct. Initialization is not enforced on fixed buffers.

tannergooding commented:
> Remove constructors from BigInteger which end up causing copy assignments that copy 464 bytes to the target variable

Is the issue that `var x = new BigInteger(...);` ends up creating a temporary and then copying the result to `x`?

If so, that seems like a codegen issue around large structs. Ideally (and if possible), the JIT would just assign directly to the result rather than incurring a copy (struct returns have to allocate space on the caller stack and then pass a pointer to that space to the callee anyways).

I actually think I logged a bug for a similar problem related to the HWIntrinsics.

CC. @CarolEidt, @AndyAyersMS. Is this already covered by the "first class structs" document?

tannergooding commented:
Looks like this is probably related to https://github.com/dotnet/coreclr/issues/16887.

AndyAyersMS commented:
> Is the issue that `var x = new BigInteger(...);` ends up creating a temporary and then copying the result to `x`?

To avoid this copy the jit has to be able to prove that `x` is completely dead before the assignment, and we're not in a position to do that in the importer. So in the importer, the JIT will assign to a temp first, then copy to the final destination. Issue tracking this is dotnet/coreclr#14890. Addressing it via reverse struct copy-prop requires being able to reason about struct liveness and hence depends on first class structs.

ts2do committed:
Add missing break

ts2do commented:
> Is the issue that `var x = new BigInteger(...);` ends up creating a temporary and then copying the result to `x`?

That's exactly right. Since the `BigInteger` struct is so large and its usage is limited to two files, it seemed the best way to avoid undesired copying was to memberwise-initialize on an out parameters.

ts2do committed:
Change ref to out in BigInteger.Multiply

Change `result` argument from `ref BigInteger` to `out BigInteger` for both overloads of Multiply

AndyAyersMS commented:
Some more notes on this.

Roslyn will detect cases like `var x = new StructType(...)` and in many cases optimize those to call the constructor directly on `&x` and avoid extra copying. But there are cases where Roslyn will fall back to `newobj`, say if the constructor takes byref params (see for instance dotnet\roslyn#16364), or if the final location for the construction is not provably on the stack.

The cases being optimized here are all cases where the final location for construction is a byref parameter, so Roslyn currently can't prove that the target is on the stack, and so emits `newobj`, and this entails construction in a temp and then copying to the final location.

Since `BigInteger` is a `ref struct` one might think Roslyn could  deduce the target must be on the stack and so enable the optimization. Likewise,  when assigning to an `out` parameter one might assume the old values at the final location could not possibly be observed during construction. But neither of these is sufficient -- there's a second condition that must be checked: the results of a partial construction can't be observed (say the constructor throws mid-way through assigning fields). To do that check, the relationship between try-nesting for the construction and the introduction of the local needs to be understood, or it needs to be proven that the constructor can't throw after assigning only some fields. Neither of these is possible, as they require some kind of interprocedural analysis.

For similar reasons the JIT will almost certainly be unable to prove that direct construction is safe in these cases, even with some of the improvements we hope to add over time. The copies we aim to eliminate in the JIT come from cases where structs are used as temporaries, things like like `F(new Struct(...))`, see dotnet/coreclr#14890 or dotnet/coreclr#14109.

So I don't see any codegen-related relief for this case, which is unfortunate, as large structs and explict byref params go hand in hand.

Selective field updates are probably the best option to avoid the copies. I'll let someone else weigh in on whether this change encapsulates those in a reasonable manner, and whether the perf benefits of the change are sufficient motivation...

ts2do committed:
Compile fix

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:
```diff
@@ -812,10 +795,6 @@ public static void Multiply(ref BigInteger lhs, ref BigInteger rhs, ref BigInteg
                 {
                     result._length = (maxResultLength - 1);
```
this could probably be simplified to `result._length -= 1` now.

ts2do replied:
Done; used -- for code style conformance.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:
```diff
@@ -1042,14 +1020,24 @@ public bool IsZero()
-                BigInteger temp = new BigInteger(0);
-                temp.SetValue(ref this);
-                Multiply(ref temp, ref value, ref this);
+                switch (value._length)
```
Is the codegen for the switch case here good? I would expect that an if statement might have better codegen and be better predicted by the CPU.

ts2do replied:
I can't compile it to test on my machine at the moment, but I would have figured switch would be fine; I don't know much about how prediction works with it.
I'm reconsidering this change because it doesn't gain us much. It will still be O(nm) regardless of the checks, the only difference being avoiding a copy. The checks inside `Multiply(ref BigInteger, ref BigInteger, out BigInteger)` should be sufficient for choosing the best way to complete the operation.

ts2do committed:
Code review change

ts2do committed:
Remove optimization from Multiply(ref BigInteger)

ts2do commented:
I've removed the optimization from `Multiply(ref BigInteger)` method, as I explained in the review comment.

Would it better to include changes not directly relating to the pull request here or should that be separated? I'm particularly interested in covering more cases in the optimization checks at the top of `Multiply(ref BigInteger, uint, out BigInteger)` and `Multiply(ref BigInteger, ref BigInteger, out BigInteger)`.

ts2do commented:
@tannergooding I've removed the optimization from `Multiply(ref BigInteger)` method, as I explained in the review comment.

Would it better to include changes not directly relating to the pull request here or should that be separated? I'm particularly interested in covering more cases in the optimization checks at the top of `Multiply(ref BigInteger, uint, out BigInteger)` and `Multiply(ref BigInteger, ref BigInteger, out BigInteger)`. Let me know, and I'll be happy to take off WIP and do whatever you recommend.

ts2do commented:
I've decided to make the unrelated changes in a subsequent PR to get this one done with.

safern commented:
@tannergooding is this good to merge?

danmosemsft commented:
ping @tannergooding 😄

mjsabby commented:
@tannergooding Any reason why this wasn't squashed and merged?

tannergooding commented:
Was an accident. GitHub looks to track the default merge state per repo and this is a new repo, so it defaults to merge.
